java/ql/src/experimental/Security/CWE/CWE-094/JexlInjectionLib.qll

```diff
@@ -0,0 +1,464 @@
+ */
+predicate creatingTaintedJexlExpression(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    (
+      m instanceof JxltEngineCreateExpressionMethod or
+      m instanceof UnifiedJexlParseMethod or
+      m instanceof JexlEngineCreateExpressionMethod
+    ) and
+    ma.getAnArgument().getType() instanceof TypeString and
+    ma.getAnArgument() = node1.asExpr() and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that creates a Jexl expression.
+ */
+predicate creatingTaintedJxltEngineExpression(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    (m instanceof JxltEngineCreateExpressionMethod or m instanceof UnifiedJexlParseMethod) and
+    ma.getAnArgument().getType() instanceof TypeString and
+    ma.getAnArgument() = node1.asExpr() and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that creates a Jexl template.
+ */
+predicate creatingTaintedJexlTemplate(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    (m instanceof JxltEngineCreateTemplateMethod or m instanceof UnifiedJexlCreateTemplateMethod) and
+    (
+      isCreateTemplateSourceArg(ma, 0, node1.asExpr()) or
+      isCreateTemplateSourceArg(ma, 1, node1.asExpr())
+    ) and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if:
+ * - `expr` is an argument with the `index`
+ * - `expr` is a string or an instance of `Reader`
+ */
+predicate isCreateTemplateSourceArg(MethodAccess ma, int index, Expr expr) {
+  (
+    ma.getArgument(index).getType() instanceof TypeString or
+    ma.getArgument(index).getType() instanceof Reader
+  ) and
+  ma.getArgument(index) = expr
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that creates a Jexl script.
+ */
+predicate creatingTaintedJexlScript(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    m instanceof JexlEngineCreateScriptMethod and
+    ma.getArgument(0).getType() instanceof TypeString and
+    ma.getArgument(0) = node1.asExpr() and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step
+ * that creates a callable from a Jexl expression or script.
+ */
+predicate creatingTaintedJexlCallable(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    (m instanceof JexlExpressionCallableMethod or m instanceof JexlScriptCallableMethod) and
+    ma.getQualifier() = node1.asExpr() and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that returns data from
+ * a tainted bean by calling one of its getters.
+ */
+predicate returningTaintedDataFromBean(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    m instanceof GetterMethod and
```

Comment:
I note we already have `m instanceof GetterMethod and m.getDeclaringType() instanceof SpringUntrustedDataType` in `taintPreservingQualifierToMethod` -- does that include the cases you need? Check if any of your tests fail without this?

Reply:
Yes, unfortunately, one of the tests fail without this (please see `Jexl3Injection.java`):

`testWithSpringControllerThatEvaluatesJexlFromRequestBodyWithNestedObjects`

This test covers the following:

- Tainted data comes from a complex bean `CustomRequest` that has `@RequestBody` annotation.
- The `CustomRequest` class contains another bean of type `Data`.

```java
    @PostMapping("/request")
    public ResponseEntity testWithSpringControllerThatEvaluatesJexlFromRequestBodyWithNestedObjects(
            @RequestBody CustomRequest customRequest) {

        String expr = customRequest.getData().getExpr();
        runJexlExpression(expr);
```

The `TaintedSpringRequestBody` flow source and `returningTaintedDataFromBean` step work together to catch this case.

I know that this step may be too broad so that it causes many false-positives (please see https://github.com/github/codeql/pull/4965#issuecomment-761885457). I've been testing this on several codebases, so far it discovered some true-positives. I still need some time to finish the analysis. Taking into account that the query is experimental, I'd like to keep this taint propagation step. The impact of a JEXL injection is arbitrary code execution. For such a high impact, maybe it's okay to sacrifice the false-positive rate a bit in order to find more true-positives.

Reply:
`TaintedSpringRequestBody` and `returningTaintedDataFromBean` allowed to detect https://github.com/traccar/traccar/issues/4624

Reply:
Do you think the code to retrieve taint from a complex bean like that could be made universal? If so let's make a separate PR for that too -- if on the other hand you think it would be too noisy and should be restricted to this JEXL query then let's keep it as you say.

Reply:
I think it would be too noisy if we make it universal (at least with the current version of the step). Let's please keep it restricted to this JEXL query.